Add support for deploy service agent auto updates

bernardjkim opened the pull request:
Contributes to https://github.com/gravitational/teleport/issues/28780, https://github.com/gravitational/cloud/issues/4880

This PR adds support for auto updates of Deploy Service Agents. More info in [RFD](https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md#aws-oidc---deploy-service-agents-amazon-ecs-with-fargate).
Some changes from the RFD:
- The RFD described the upgrade process by first calling `ecs:CreateService` and then `ecs:DeleteService`, to update a service, but the implementation just uses `ecs:UpdateService` to update the service.

bernardjkim committed:
Add support for ecs agent auto updates

bernardjkim committed:
fix unit test

bernardjkim commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,284 @@
+
+// AWSRegionsList is the list of available AWS regions
+// https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html
+var AWSRegionsList = []string{
```
The current implementation iterates over all regions to check for available Deploy Service Agents. Anyone have an idea for a more efficient way of finding all available Deploy Service Agents?

r0mant replied:
Yes, we should iterate over all registered databases to get a list of regions.

bernardjkim replied:
> To query for available ECS clusters Teleport will iterate over all databases filtering for RDS, and then extract the AWS.Region.

I was thinking this would require the same number of API calls because we'd need to query each region individually for rds databases. Is this information cached by Teleport and available for lookup?

marcoandredinis replied:
It should be cached, yes.

You will iterate over all Teleport Databases and for those that are `AWS`, extract its region.

This won't give you all the DatabaseService agents.
It is possible to have deployed an agent and having 0 databases.
I think we are ok with this.

bernardjkim committed:
Remove unused var

r0mant commented on lib/automaticupgrades/version.go:
```diff
@@ -73,3 +72,60 @@ func Version(ctx context.Context, baseURL string) (string, error) {
+}
+
+// Critical returns true if a critical upgrade is available.
+func Critical(ctx context.Context, criticalURL string) (bool, error) {
```
This is the same as `Version()`, just fetches a different endpoint, let's factor out common logic into a separate method that both `Version()` and `Critical()` can use.

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,245 @@
+
+	// Acquire the semaphore before attempting to update the deploy service agents.
+	// This task should only run on a single instance at a time.
+	lock, err := services.AcquireSemaphoreWithRetry(process.GracefulExitContext(),
```
Hmm not sure I understand the purpose of this semaphore: do we acquire the lock before starting the service? Shouldn't we acquire it only when actually running the upgrade? Cause the semaphore will expire after updateDeployAgentsInterval and the 2nd process will start anyway?

bernardjkim replied:
It seems like the semaphore continues to renew as long as the process continues to run. So the 2nd process will only be able to acquire the lease after the 1st process is terminated. So there is only on instance of this task running at a time.

I'll read through the code again and make necessary changes. I'm not entirely confident I'm using the semaphores correctly here.

bernardjkim replied:
I've made some changes to how I was using the semaphore locks. Let me know if that looks correct.

r0mant reviewed (commented):
First quick pass, I'll do another more in-depth one later.

smallinsky commented on lib/auth/integration/integrationv1/awsoidc.go:
```diff
@@ -41,42 +45,89 @@ func (s *Service) GenerateAWSOIDCToken(ctx context.Context, req *integrationpb.G
+	return nil
+}
+
+// GenerateToken generates a token to be used when executing an AWS OIDC Integration action.
+func GenerateAWSOIDCToken(ctx context.Context, config AWSOIDCTokenConfig) (string, error) {
```
```suggestion
// GenerateAWSOIDCToken generates a token to be used when executing an AWS OIDC Integration action.
func GenerateAWSOIDCToken(ctx context.Context, config AWSOIDCTokenConfig) (string, error) {
```

smallinsky commented on lib/automaticupgrades/version.go:
```diff
@@ -73,3 +72,60 @@ func Version(ctx context.Context, baseURL string) (string, error) {
+
+	critical := strings.TrimSpace(string(body))
+
+	// The critical endpoint returns either the string "yes" or "no"
+	return critical == "yes", nil
```
The body can have different payload can we handled these value separate and returns a error if  value is different that yes or no ?

smallinsky commented on lib/automaticupgrades/version.go:
```diff
@@ -73,3 +72,60 @@ func Version(ctx context.Context, baseURL string) (string, error) {
+		return false, trace.Wrap(err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
```
Do we plan to support HTTPS proxy ?

bernardjkim replied:
This topic has not been discussed for this task. Do you think we need support in this PR. Otherwise we can create an issue and implement in a future PR.

smallinsky commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer process.GetAuthServer().CancelSemaphoreLease(process.GracefulExitContext(), *lock)
```
nit: can we handle the error ? 
```suggestion
	defer func() { 
	if err := process.GetAuthServer().CancelSemaphoreLease(process.GracefulExitContext(), *lock); err != nil {
     s.log.WithError(err).Errorf("Failed to cancel lease: %v.", lease)
	}
}()
```

bernardjkim commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,245 @@
+)
+
+func (process *TeleportProcess) periodUpdateDeployServiceAgents() error {
+	if !process.Config.Auth.Enabled {
+		return nil
+	}
```
I decided to run this in the auth instead of proxy for two reasons

- The proxy service does not have permissions to read the cluster_maintenance_config that contains the upgrade window configuration.
- The proxy services are not configured with the TELEPORT_AUTOMATIC_UPGRADES env variable.

If you feel strongly that this task should run in the proxy we can make the necessary changes. But it is a bit simpler to run the task on auth.

marcoandredinis replied:
I really think we should do this within the proxy.
Auth should be reserved for more critical tasks (eg generating certs and tokens).

Adding read permissions for `cluster_maintenance_config` should be ok.
The `TELEPORT_AUTOMATIC_UPGRADES` should be available in the `.../webapi/ping` endpoint, so I think Proxy should already have access to that value.

What do you think @r0mant ?

bernardjkim replied:
> Auth should be reserved for more critical tasks (eg generating certs and tokens).

That's a valid concern. I've switch the task to run on proxies.

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Addres feedback

bernardjkim committed:
Use list of available AWS database regions

bernardjkim committed:
Run update task on proxy instances

bernardjkim committed:
Revert GenerateAWSOIDCToken

bernardjkim committed:
Move const to start of file

marcoandredinis commented on lib/cloud/aws/policy_statements.go:
```diff
@@ -50,7 +50,7 @@ func StatementForECSManageService() *Statement {
 			"ecs:DescribeClusters", "ecs:CreateCluster", "ecs:PutClusterCapacityProviders",
 			"ecs:DescribeServices", "ecs:CreateService", "ecs:UpdateService",
-			"ecs:RegisterTaskDefinition",
+			"ecs:RegisterTaskDefinition", "ecs:DescribeTaskDefinition", "ecs:DeregisterTaskDefinition",
```
We have to think of some way to let the users know that their current set of permissions might not be enough, and that they must re-run the script.

Maybe a warning in the logs?
Cluster Alert might also work

marcoandredinis replied:
Are we doing anything here?

For new deployments, it should be good (we add all the permissions).
For current deployments, it will fail.
Do we have any kind of alerting?

bernardjkim replied:
We will log the following message if permissions are lacking, but I don't know how effective that will be. https://github.com/gravitational/teleport/pull/31982/files#diff-b2b691cbb3f300d6396f5b1ff26ac2569a944859ac8041067d8c97b23037f9c3R321-R325

I'm not familiar with Cluster Alerts, do you think this would be a better approach?

marcoandredinis replied:
The log messages are stored internally (as in Teleport Cloud infra).
They will not be visible to the customer.

Maybe we can get away with looking for those logs ourselves and letting the affected customers know.
But honestly, that feels hacky, error-prone and not scalable.

With a ClusterAlert the customer would know.
I'm not very familiar with those as well, but I believe it would work better (even if we do this, let's keep the warning as well).
@r0mant What do you think?


As for method to fix:
> Re-run deploy service configuration script to update permissions.

We don't have an easy way for the user to re-run the script.
It is only generated at that step when enrolling an RDS database.

Can we generate the script URL?
It should look something like this:

```
curl 'https://<tenant>.teleport.sh/webapi/scripts/integrations/configure/deployservice-iam.sh?integrationName=<integration-name>&awsRegion=<aws-region>&taskRole=<taskRole>&role=<integrationRole>'
```

I think all the variables are present in the current context.

bernardjkim replied:
I've added a cluster alert here [fec081d](https://github.com/gravitational/teleport/pull/31982/commits/fec081d114770b57bd29ec4d4acf5bfde990b3fd)

Do you know of a way to get the task role in the current context? The only way I'm aware of is extracting the task role from the task definition. But that would require that the instance already has permissions to describe the task definition.

marcoandredinis replied:
Yeah, I don't think we can.
I thought it was under Service, but it is actually under Task Definition 😭 

In that case, we can only ask them to fill it with the same role that was used previously.

bernardjkim replied:
Okay the alert might look something like the following:
```
Open Amazon CloudShell and copy/paste the following command to reconfigure integration. Replace TASK_ROLE with deploy service task role name. bash -c "$(curl 'https://bernard-dev.cloud.gravitational.io/webapi/scripts/integrations/configure/deployservice-iam.sh?awsRegion=us-west-2&integrationName=database-access&role=bernard-dev-database-access&taskRole=TASK_ROLE')"
```

It is pretty verbose tho. We might want to consider adding a docs page later and linking that instead.

bernardjkim replied:
Let's leave cluster alerts out of scope for now and revisit at a later time.

marcoandredinis commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,220 @@
+
+// UpdateDeployServiceAgents updates the deploy service agents with the specified teleportVersionTag.
+func UpdateDeployServiceAgents(ctx context.Context, clt DeployServiceClient, teleportClusterName, teleportVersionTag string, ownershipTags AWSTags) error {
+	teleportFlavor := teleportOSS
+	if modules.GetModules().BuildType() == modules.BuildEnterprise {
+		teleportFlavor = teleportEnt
+	}
+	teleportImage := fmt.Sprintf("public.ecr.aws/gravitational/%s-distroless:%s", teleportFlavor, teleportVersionTag)
```
Can we move this to a method and re-use it for the `upsertTask` ?

marcoandredinis commented on lib/service/service.go:
```diff
@@ -1194,6 +1194,8 @@ func NewTeleport(cfg *servicecfg.Config) (*TeleportProcess, error) {
 	// at any time with dynamic configuration
 	process.RegisterFunc("common.upload.init", process.initUploaderService)
 
+	process.RegisterFunc("update.deploy.agents", process.periodUpdateDeployServiceAgents)
+
```
Can we add `aws-oidc` or something related?
So that it's easier to understand the context just from the method name

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+
+// withinUpgradeWindow returns true if the current time is within the configured
+// upgrade window.
+func withinUpgradeWindow(cmc types.ClusterMaintenanceConfig, clock clockwork.Clock) bool {
+	upgradeWindow, ok := cmc.GetAgentUpgradeWindow()
+	if !ok {
+		return false
+	}
+
+	now := clock.Now()
+	if len(upgradeWindow.Weekdays) == 0 {
+		if int(upgradeWindow.UTCStartHour) == now.Hour() {
+			return true
+		}
+	}
+
+	weekday := now.Weekday().String()
+	for _, upgradeWeekday := range upgradeWindow.Weekdays {
+		if weekday == upgradeWeekday {
+			if int(upgradeWindow.UTCStartHour) == now.Hour() {
+				return true
+			}
+		}
+	}
+	return false
+}
```
I think this method belongs to something more generic :thinking: 
Maybe `lib/automaticupgrades` ?

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+
+			// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+			// If the lease cannot be acquired, the update is already being handled by another instance.
+			semLock, err := authClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
```
It looks like the semaphore is only released explicitly in case of error below. Should we release it in case of success as well?

bernardjkim replied:
If we release the semaphore on success, other instances can acquire the lock in the same update interval. This means that the update may be attempted for the same Deploy Service Agents by multiple instances. This isn't necessarily a problem because the update will return early after comparing the current Teleport version and the stable version, but it does make a few API calls before then. So holding onto the semaphore until the next update interval will help us reduce the number of API calls made.

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+}
+
+// getStableTeleportVersion returns the current stable version of teleport
+func getStableTeleportVersion(ctx context.Context) (string, error) {
```
Similar to @marcoandredinis's comment below this method doesn't seem to belong in the "awsoidc" module. Perhaps lib/automaticupgrades would be a better place for it?

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+		}
+		integrationName := ig.GetName()
+
+		for region := range awsRegions {
```
It looks like we're iterating over every integration times every region. Do we have information about which database was discovered by which integration so we can only iterate over regions that actually have databases for a particular integration?

bernardjkim replied:
Not that I know of. @marcoandredinis?

marcoandredinis replied:
No, we don't include anything that could backtrack the integration used.
I don't expect people to have a lot of integrations, but I still think we should improve here.

What do you think about adding a label?
It might bring false positives, because they might have a self-hosted DatabaseService proxying the connection and not using the ECS Agent.
Still an improvement 👍 
Let me know what you think

bernardjkim replied:
> What do you think about adding a label?

Would this mean adding a label to the database?
Is this something that should be included in this PR?

marcoandredinis replied:
No, I don't think it is worth it.
Let's keep the current scope

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
```
This function needs comments explaining key bits of business logic in relevant parts of the function. For example, how we're collecting regions, that we're using OIDC integration clients, etc.

I know it's technically all in the code but it may be confusing for new readers and comments really help with code comprehension.

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+
+	var resources []types.Integration
+	var nextKey string
+	for {
```
Do we not already have a helper method that returns all integrations? If not, can we make it somewhere appropriate?

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+				return trace.Wrap(err)
+			}
+
+			if err := awsoidc.UpdateDeployServiceAgents(ctx, deployServiceClient, clusterNameConfig.GetClusterName(), teleportVersion, ownershipTags); err != nil {
```
Will there be anything in the logs indicating the service is attempting an auto-upgrade?

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,263 @@
+	}
+}
+
+func (process *TeleportProcess) updateDeployServiceAgents(ctx context.Context, authClient *auth.Client) error {
```
I don't know if this should be a part of `TeleportProcess` to be honest.

Can we make a proper separate "service struct" e.g. `FargateUpdater` or something in lib/automaticupgrades, that would get initialized with a config struct, implement all necessary upgrade business logic (basically, this method) and provide a proper interface to Teleport process to call?

r0mant commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,220 @@
+const waitDuration = time.Minute * 5
+
+// UpdateDeployServiceAgents updates the deploy service agents with the specified teleportVersionTag.
+func UpdateDeployServiceAgents(ctx context.Context, clt DeployServiceClient, teleportClusterName, teleportVersionTag string, ownershipTags AWSTags) error {
```
Would it make sense to make this (and `updateDeployServiceAgent` below) a member of `DeployServiceClient`? E.g. `clt.UpdateAgents()`.

bernardjkim replied:
I don't think `UpdateDeployServiceAgent` should be a member of the interface, but I'll add `UpdateServiceRequest` struct to simplify the function def.

r0mant commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,220 @@
+	_, err = updateServiceOrRollback(ctx, clt, service, registerTaskDefinitionOut.TaskDefinition)
+	if err != nil {
+		// If update failed, then rollback task definition
+		_, rollbackErr := clt.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
```
What happens after rollback? Will we reattempt the upgrade again?

bernardjkim replied:
If the update fails, it will be reattempted during the next interval (30mins), but only if it is still within the upgrade window or the critical upgrade flag is still enabled.

r0mant commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,220 @@
+}
+
+func getManagedService(ctx context.Context, clt DeployServiceClient, teleportClusterName string, ownershipTags AWSTags) (*ecsTypes.Service, error) {
+	ecsClusterName := fmt.Sprintf("%s-teleport", normalizeECSResourceName(teleportClusterName))
```
Is there a function/constant for this? Feels like this is used in more than a single place.

r0mant reviewed (commented):
I think the overall approach makes sense but I left a few suggestions about code structure/organization.

bernardjkim committed:
Address feedback

bernardjkim commented on lib/automaticupgrades/version.go:
```diff
@@ -34,22 +34,54 @@ const (
+
+	// Expects critical endpoint to return either the string "yes" or "no"
+	switch critical {
+	case "yes":
```
Not sure why we decided on "yes"/"no" for the critical flag value, but they are not considered valid boolean values.

bernardjkim commented on lib/integrations/awsoidc/deployservice.go:
```diff
@@ -456,11 +464,16 @@ func upsertTask(ctx context.Context, clt DeployServiceClient, req DeployServiceR
 				configB64,
 			},
-			EntryPoint: []string{"teleport"},
+			EntryPoint: []string{"/usr/bin/dumb-init"},
```
Verified that `dumb-init` is available in the distroless image.

bernardjkim committed:
Create separate DeployServiceUpdater struct

marcoandredinis commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,233 @@
+			TaskDefinition: registerTaskDefinitionOut.TaskDefinition.TaskDefinitionArn,
+		})
+		if rollbackErr != nil {
+			return trace.Wrap(err, "failed to rollback task definition: %v", rollbackErr)
```
We hide the `updateServiceOrRollback` error
Can we instead use a `trace.Aggregate` error?

marcoandredinis commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	// Attempt to deregister previous task definition but ignore error on failure
+	clt.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
+		TaskDefinition: taskDefinition.TaskDefinitionArn,
+	})
```
Can we please log the error?

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,301 @@
+	}
+
+	updater, err := NewDeployServiceUpdater(DeployServiceUpdaterConfig{
+		Log:                 process.log.WithField(trace.Component, teleport.Component(teleport.ComponentProxy, "deployserviceupdater")),
```
```suggestion
		Log:                 process.log.WithField(trace.Component, teleport.Component(teleport.ComponentProxy, "awsoidc-deployserviceupdater")),
```
Or similar.

I really think we should mention the `aws-oidc`
Because deploy service are two very generic terms.

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,301 @@
+				TeleportVersionTag:  cloudStableVersion,
+				OwnershipTags:       ownershipTags,
+			}); err != nil {
+				if trace.IsNotFound(err) {
+					// The updater checks each integration/region combination, so
+					// there will be regions where there is no ECS cluster deployed
+					// for the integration
+					continue
+				}
```
Can we log this as a debug message?

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,301 @@
+			}
+
+			updater.Log.Debugf("Updating Deploy Service Agents in AWS region: %s", region)
+			if err := awsoidc.UpdateDeployServiceAgent(ctx, deployServiceClient, awsoidc.UpdateServiceRequest{
```
Given that we wait for the ECS service to be ready, I think we should parallelize per Integration/Region.
A waitgroup with maybe 3 slots should be ok (I don't think we should have unbounded parallelization)

marcoandredinis reviewed (commented):
Just tested this (very basic test) and it worked as expected 👍 

![image](https://github.com/gravitational/teleport/assets/689271/f4928e47-b350-485f-ac36-39cb1a204852)

Left some comments

bernardjkim committed:
Address feedback

- Perform updates in parallel
- Add additional logging
- Add additional documentation

bernardjkim committed:
debug

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,330 @@
+			wg.Add(1)
+			go func(ig types.Integration, region string) {
+				if err := updater.updateDeployServiceAgent(ctx, ig, region, stableVersion); err != nil {
+					updater.Log.WithError(err).Warning("Failed to update deploy service agent.")
```
Log integration name and region?

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	// Perform updates in parallel across regions.
+	var sem = make(chan interface{}, maxConcurrentUpdates)
+	var wg sync.WaitGroup
```
:nail_care: 
Maybe we should use
https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool instead

Sorry for recommending a WaitGroup

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,330 @@
+		}
+		return trace.Wrap(err)
+	}
+
```
Is there a reason why we only call the `CancelSemaphoreLease` on error?

I would expect that we would call `defer ... CancelSemaphoreLease` here

bernardjkim replied:
If we hold on to the semaphore lease, other instances will not be able to acquire the lease within the same 30min interval. This will reduce the number of API calls a bit.

A small optimization, but could potentially be prone to bugs in future code changes. I'll go ahead and just cancel the semaphore on success as well.

marcoandredinis commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (req *UpdateServiceRequest) CheckAndSetDefaults() error {
```
I think we should check for the presence of ownership tags as well.

bernardjkim committed:
Address feedback

- Check OwnershipTags
- Use semaphore pkg
- Release semaphore lease on success

bernardjkim commented:
Thanks for all the great feedback!

bernardjkim committed:
Make OwnershipTags explicitly required

camscale commented:
I've removed myself from the reviewers as you already have three with some great comments. Rather than trying to build up context myself, I'll leave it in their capable hands for approval.

bernardjkim committed:
Add cluster alert

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -323,8 +330,84 @@ func (updater *DeployServiceUpdater) updateDeployServiceAgent(ctx context.Contex
+	values.Add("integrationName", integration.GetName())
+	values.Add("awsRegion", awsRegion)
+	values.Add("taskRole", url.QueryEscape("<task-role>")) // The task role needs to be supplied by the user
+	values.Add("taskRole", path.Base(integration.GetAWSOIDCIntegrationSpec().RoleARN))
```
This one should be `role`

marcoandredinis commented on lib/service/awsoidc.go:
```diff
@@ -323,8 +330,84 @@ func (updater *DeployServiceUpdater) updateDeployServiceAgent(ctx context.Contex
+	values := scriptURL.Query()
+	values.Add("integrationName", integration.GetName())
+	values.Add("awsRegion", awsRegion)
+	values.Add("taskRole", url.QueryEscape("<task-role>")) // The task role needs to be supplied by the user
```
I think we can get this from the ECS Service (which we should have access to even for existing clusters)

bernardjkim committed:
Fix typo and update message

r0mant commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,244 @@
+	// then rollback service with previous task definition
+	rollbackServiceOut, rollbackErr := clt.UpdateService(ctx, generateServiceWithTaskDefinition(service, aws.ToString(service.TaskDefinition)))
+	if rollbackErr != nil {
+		return nil, trace.Wrap(err, "failed to rollback service: %v", err)
```
Should probably use `trace.NewAggregate` to include both `err` and `rollbackErr` here.

r0mant commented on lib/integrations/awsoidc/deployservice_update.go:
```diff
@@ -0,0 +1,244 @@
+		Cluster:  updateServiceOut.Service.ClusterArn,
+	}, waitDuration)
+	if rollbackErr != nil {
+		return nil, trace.Wrap(err, "failed to rollback service: %v", err)
```
Same as above, include `rollbackErr` as well.

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,414 @@
+			// The updater checks each integration/region combination, so
+			// there will be regions where there is no ECS cluster deployed
+			// for the integration.
+			updater.Log.WithError(err).Debugf("Integration %s does not manage any services within region %s.", integration.GetName(), awsRegion)
```
Nit: It will be very verbose with the error trace.
```suggestion
			updater.Log.Debugf("Integration %s does not manage any services within region %s.", integration.GetName(), awsRegion)
```

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,414 @@
+			// In this situation users should be notified that they will need to
+			// re-run the deploy service iam configuration script and update the
+			// permissions.
+			updater.Log.WithError(err).Warning("Re-run deploy service configuration script to update permissions.")
```
Can we make this error message more useful? If I saw it, I would have no idea what it means and what I need to do.

bernardjkim replied:
I've simplified the message to "Update integration role and add missing permissions.". The error already includes context on relevant role and operation 
```
error ECS: DescribeTaskDefinition, https response error StatusCode: 400, RequestID: 57f3de8c-3f42-4081-a004-d789d41f6475, api error AccessDeniedException: User: arn:aws:sts::278576220453:assumed-role/bernard-dev-database-access/1694123674929138369 is not authorized to perform: ecs:DescribeTaskDefinition on resource: * because no identity-based policy allows the ecs:DescribeTaskDefinition action
```

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,414 @@
+
+	alertLabels := map[string]string{
+		types.AlertAWSOIDCAccessDenied: "yes",
+		types.AlertOnLogin:             "yes",
```
I don't know if we need to alert on login. Every user will see it, and most of them won't be able to do anything about it.

r0mant commented on lib/service/awsoidc.go:
```diff
@@ -0,0 +1,414 @@
+
+// ensureClusterAlert ensures a cluster alert is created if deploy service permissions
+// need to be reconfigured.
+func (updater *DeployServiceUpdater) ensureClusterAlert(ctx context.Context, integration types.Integration, awsRegion string) error {
```
I'm not sure a cluster alert is an appropriate tool for this kind of notification. I think we need to come up with a consistent way of how to attract appropriate users' attention to actions they need to perform, but that's out of scope of this issue.

I would keep cluster alerts out of the initial implementation, and I'll bring it up with the product team to see what we want to do going forward.

bernardjkim committed:
Revert cluster alert

bernardjkim committed:
Update err messages

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Merge branch 'master' into bernard/ecs-auto-update

bernardjkim committed:
Check minimum compatible server version

bernardjkim committed:
Update log msg
